pkg/steps/censoringclient/client.go

```diff
@@ -0,0 +1,138 @@
+package censoringclient
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	"github.com/openshift/ci-tools/pkg/secrets"
+	"github.com/openshift/ci-tools/pkg/util/watchingclient"
+)
+
+func Wrap(upstream watchingclient.Client, censor *secrets.DynamicCensor) watchingclient.Client {
+	return &client{
+		upstream: upstream,
+		censor:   censor,
+	}
+}
+
+// client updates the censor when it interacts with secrets on the cluster
+type client struct {
+	upstream watchingclient.Client
+	censor   *secrets.DynamicCensor
+}
+
+func (c *client) Get(ctx context.Context, key ctrlruntimeclient.ObjectKey, obj ctrlruntimeclient.Object) error {
+	if err := c.upstream.Get(ctx, key, obj); err != nil {
+		return err
+	}
+	c.record(obj)
+	return nil
+}
+
+func (c *client) List(ctx context.Context, list ctrlruntimeclient.ObjectList, opts ...ctrlruntimeclient.ListOption) error {
+	if err := c.upstream.List(ctx, list, opts...); err != nil {
+		return err
+	}
+	c.recordList(list)
+	return nil
+}
+
+func (c *client) Create(ctx context.Context, obj ctrlruntimeclient.Object, opts ...ctrlruntimeclient.CreateOption) error {
+	if err := c.upstream.Create(ctx, obj, opts...); err != nil {
+		return err
+	}
+	c.record(obj)
+	return nil
+}
+
+func (c *client) Delete(ctx context.Context, obj ctrlruntimeclient.Object, opts ...ctrlruntimeclient.DeleteOption) error {
+	return c.upstream.Delete(ctx, obj, opts...)
+}
+
+func (c *client) Update(ctx context.Context, obj ctrlruntimeclient.Object, opts ...ctrlruntimeclient.UpdateOption) error {
+	if err := c.upstream.Update(ctx, obj, opts...); err != nil {
+		return err
+	}
+	c.record(obj)
+	return nil
+}
+
+func (c *client) Patch(ctx context.Context, obj ctrlruntimeclient.Object, patch ctrlruntimeclient.Patch, opts ...ctrlruntimeclient.PatchOption) error {
+	if err := c.upstream.Patch(ctx, obj, patch, opts...); err != nil {
+		return err
+	}
+	c.record(obj)
+	return nil
+}
+
+func (c *client) DeleteAllOf(ctx context.Context, obj ctrlruntimeclient.Object, opts ...ctrlruntimeclient.DeleteAllOfOption) error {
+	return c.upstream.DeleteAllOf(ctx, obj, opts...)
+}
+
+func (c *client) Watch(ctx context.Context, obj ctrlruntimeclient.ObjectList, opts ...ctrlruntimeclient.ListOption) (watch.Interface, error) {
+	return c.upstream.Watch(ctx, obj, opts...)
+}
+
+func (c *client) recordList(obj ctrlruntimeclient.ObjectList) {
```

Comment:
@deads2k could you please review from this line down?

Reply:
The `apiutil.GVKForObjece` dance is not wrong but not needed, because the client will stop working entirely if the same go type is registered in more than one api group, so just type asserting is enough

Reply:
IIRC David said if it was `unstructured.Unstructured` we needed it?

Reply:
Well the code doesn't handle unstructured.Unstructured in any way, as it will just return if the object isn't a `Secret{,List}`. It seems a tad overkill to make it handle that to me, because I doubt we will ever request a Secret through an unstructured. If we really want to handle that case, we can just type assert to unstructured.Unstructured as well and then use the TypeMeta to check if this is for a Secret.

Reply:
Makes sense - I do not understand in what cases we'd be getting one of those. Would we have to opt into it somehow on the client side?

Reply:
well, if our code did something like
```
u := &unstructured.Unstructured{}
u.SetApiVersion("v1")
u.SetKind("Secret")
client.Get(ctx, someSecretName, u)
```

But I don't see any reason why we would do that instead of using a Secret directly.

Reply:
Yeah, we can enforce that in review.